python/ray/data/preprocessors/discretizer.py

```diff
@@ -0,0 +1,360 @@
+from typing import Iterable, List, Dict, Optional, Type, Union
+
+import pandas as pd
+import numpy as np
+
+from ray.data import Dataset
+from ray.data.aggregate import Max, Min
+from ray.data.preprocessor import Preprocessor
+
+
+class _AbstractKBinsDiscretizer(Preprocessor):
+    """Abstract base class for all KBinsDiscretizers.
+
+    Essentially a thin wraper around ``pd.cut``.
+
+    Expects either ``self.stats_`` or ``self.bins`` to be set and
+    contain {column:list_of_bin_intervals}.
+    """
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        def bin_values(s: pd.Series) -> pd.Series:
+            labels = self.dtypes.get(s.name) if self.dtypes else False
+            ordered = True
+            if labels:
+                if isinstance(labels, pd.CategoricalDtype):
+                    ordered = labels.ordered
+                    labels = list(labels.categories)
+                else:
+                    labels = False
+
+            bins = self.stats_ if self._is_fittable else self.bins
+            return pd.cut(
+                s,
+                bins[s.name] if isinstance(bins, dict) else bins,
+                right=self.right,
+                labels=labels,
+                ordered=ordered,
+                retbins=False,
+                include_lowest=self.include_lowest,
+                duplicates=self.duplicates,
+            )
+
+        return df.apply(bin_values, axis=0)
+
+    def _validate_bins_columns(self):
+        if isinstance(self.bins, dict) and not all(
+            col in self.bins for col in self.columns
+        ):
+            raise ValueError(
+                "If `bins` is a dictionary, all elements of `columns` must be present "
+                "in it."
+            )
+
+    def __repr__(self):
+        attr_str = ", ".join(
+            [
+                f"{attr_name}={attr_value!r}"
+                for attr_name, attr_value in vars(self).items()
+            ]
+        )
+        return f"{self.__class__.__name__}({attr_str})"
+
+
+class CustomKBinsDiscretizer(_AbstractKBinsDiscretizer):
+    """Bin values into discrete intervals using custom bin edges.
+
+    Columns must contain numerical values.
+
+    Examples:
+        Use :class:`CustomKBinsDiscretizer` to bin continuous features.
+
+        >>> import pandas as pd
+        >>> import ray
+        >>> from ray.data.preprocessors import CustomKBinsDiscretizer
+        >>> df = pd.DataFrame({
+        ...     "value_1": [0.2, 1.4, 2.5, 6.2, 9.7, 2.1],
+        ...     "value_2": [10, 15, 13, 12, 23, 25],
+        ... })
+        >>> ds = ray.data.from_pandas(df)  # doctest: +SKIP
+        >>> discretizer = CustomKBinsDiscretizer(
+        ...     columns=["value_1", "value_2"],
+        ...     bins=[0, 1, 4, 10, 25]
+        ... )
+        >>> discretizer.transform(ds).to_pandas()  # doctest: +SKIP
```

Comment:
Can we remove `# doctest: +SKIP` now that https://github.com/ray-project/ray/pull/28395 is merged, or does it still cause problems?